### PR TITLE
fix(syslogexporter): validate all inputs at once and add display all error messages

### PR DESCRIPTION
**Pull request by rnishtala-sumo**
fix(syslogexporter): validate all inputs at once and add display all error messages

### CODE REVIEW

File: pkg/exporter/syslogexporter/config.go [hunk 1]
```diff
@@ -24,10 +24,10 @@ import (
...
 
 var (
-	unsupportedPort     = errors.New("Unsupported port: port is required, must be in the range 1-65535")
-	invalidEndpoint     = errors.New("Invalid endpoint: endpoint is required, must be a valid FQDN or IP address")
-	unsupportedProtocol = errors.New("Unsupported protocol: protocol is required, only tcp/udp supported")
-	unsupportedFormat   = errors.New("Unsupported format: Only rfc5424 and rfc3164 supported")
+	unsupportedPort     = "Unsupported port: port is required, must be in the range 1-65535"
+	invalidEndpoint     = "Invalid endpoint: endpoint is required, must be a valid FQDN or IP address"
+	unsupportedProtocol = "Unsupported protocol: protocol is required, only tcp/udp supported"
+	unsupportedFormat   = "Unsupported format: Only rfc5424 and rfc3164 supported"
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/config.go:30]: By convention errors are lower case :shrug:

```suggestion
	unsupportedPort     = "unsupported port: port is required, must be in the range 1-65535"
	invalidEndpoint     = "invalid endpoint: endpoint is required, must be a valid FQDN or IP address"
	unsupportedProtocol = "unsupported protocol: protocol is required, only tcp/udp supported"
	unsupportedFormat   = "unsupported format: Only rfc5424 and rfc3164 supported"
```

File: pkg/exporter/syslogexporter/config.go [hunk 2]
```diff
@@ -54,23 +54,28 @@ type Config struct {
...
 	switch cfg.Format {
 	case formatRFC3164Str:
 	case formatRFC5424Str:
 	default:
-		return unsupportedFormat
+		invalidFields = append(invalidFields, unsupportedFormat)
+	}
+
+	if len(invalidFields) > 0 {
+		return fmt.Errorf("%s", strings.Join(invalidFields, "\n"))
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/config.go:78]: Please use `,` as separator
```suggestion
		return fmt.Errorf("%s", strings.Join(invalidFields, ","))
```

**Review by sumo-drosiek — COMMENTED** [review]
I don't remember if it is used in Otelcol, but there is go function to join multiple errors into one: https://pkg.go.dev/errors#Join

File: pkg/exporter/syslogexporter/exporter.go [hunk 3]
```diff
@@ -121,32 +125,47 @@ func (se *syslogexporter) getTimestamp(record plog.LogRecord) time.Time {
...
-	s, err := Connect(se.logger, se.config, se.tlsConfig, se.hostname, se.pid, se.app)
-	if err != nil {
-		return fmt.Errorf("error connecting to syslog server: %s", err)
+	var errs []error
+	s, errConn := Connect(se.logger, se.config, se.tlsConfig, se.hostname, se.pid, se.app)
+	if errConn != nil {
+		errs = append(errs, errConn)
+		errStr := fmt.Sprintf("Unable to send syslog messages: %s", strings.Join(errorListToStringSlice(errs), ", "))
+		return errors.New(errStr)
 	}
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/exporter.go:134]: We should return `consumererror` which contains unsent logs. Please see the following exmaples in sumoplogicexporter:

- https://github.com/SumoLogic/sumologic-otel-collector/blob/dc1c3185755b21b2252a8f7d01a5a596e1134614/pkg/exporter/sumologicexporter/exporter.go#L180-L182
- https://github.com/SumoLogic/sumologic-otel-collector/blob/dc1c3185755b21b2252a8f7d01a5a596e1134614/pkg/exporter/sumologicexporter/exporter.go#L250-L253

Comment (kkujawa-sumo) [reply pkg/exporter/syslogexporter/exporter.go:134]: but in syslog exporter log messages may contain token which is required by sumologic ;/

Comment (sumo-drosiek) [reply pkg/exporter/syslogexporter/exporter.go:134]: Good point, but I don't think that returned logs are printed anywhere

Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/exporter.go:134]: so what should we do here. This was a deliberate choice after discussing with Kasia. This commit focuses more on error handling.

Comment (sumo-drosiek) [reply pkg/exporter/syslogexporter/exporter.go:134]: it should return: `return consumererror.NewLogs(multierr.Combine(errs...), ld)`

where `ld` contains unsent logs

Comment (sumo-drosiek) [reply pkg/exporter/syslogexporter/exporter.go:134]: and errs is a list of errors to print for customer

Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/exporter.go]: pushed changes to support `consumererror` with dropped logs

Comment (kkujawa-sumo) [reply pkg/exporter/syslogexporter/exporter.go]: @rnishtala-sumo Could you check if returned logs are printed, see: https://github.com/SumoLogic/sumologic-otel-collector/pull/986#discussion_r1108339762?

Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/exporter.go]: it doesn't look like it
```2023-02-16T11:29:33.070-0500	error	exporterhelper/queued_retry.go:303	Exporting failed. Dropping data. Try enabling sending_queue to survive temporary failures.	{"kind": "exporter", "data_type": "logs", "name": "syslog", "dropped_items": 1}```

**Comment by rnishtala-sumo** [comment]
looks like we hit a limit in one of the tests, otherwise the PR checks seem fine
```<?xml version="1.0" encoding="utf-8"?><Error><Code>ServerBusy</Code><Message>Egress is over the account limit.```

File: pkg/exporter/syslogexporter/config.go [hunk 4]
```diff
@@ -52,23 +52,28 @@ type Config struct {
...
 	switch cfg.Format {
 	case formatRFC3164Str:
 	case formatRFC5424Str:
 	default:
-		return unsupportedFormat
+		invalidFields = append(invalidFields, unsupportedFormat)
+	}
+
+	if len(invalidFields) > 0 {
+		return fmt.Errorf("%s", strings.Join(invalidFields, "; "))
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/config.go:76]: There is `errors.Join` function which could do the trick, but for some reason it is not used in otelcol code at all :confused: Anyway after few days of thinking about it I think we should use it

Comment (sumo-drosiek) [reply pkg/exporter/syslogexporter/config.go:76]: I know why. They are using `multierr.Combine` :D

Comment (sumo-drosiek) [reply pkg/exporter/syslogexporter/config.go:76]: Please use `invalidFields := []error` and then `return multierr.Combine(invalidFields...)`

File: pkg/exporter/syslogexporter/exporter.go [hunk 5]
```diff
@@ -102,32 +106,71 @@ func (se *syslogexporter) getTimestamp(record plog.LogRecord) time.Time {
...
+		}
+	}
+	if len(dropped) > 0 {
+		ld = plog.NewLogs()
+		for i := range dropped {
+			rls := ld.ResourceLogs().AppendEmpty()
+			logRecords := rls.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+			dropped[i].resource.MoveTo(rls.Resource())
+			for j := 0; j < len(dropped[i].records); j++ {
+				dropped[i].records[j].MoveTo(logRecords)
```
Comment (sumo-drosiek) [inline pkg/exporter/syslogexporter/exporter.go:135]: Please add unit tests for this

Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/exporter.go:135]: do we need a test syslog server in our tests (mock) to write tests for the `pushLogsData`?

File: pkg/exporter/syslogexporter/config.go (hunk 4 above)
Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/config.go]: I made this change

**Review by sumo-drosiek — APPROVED** [review]
LGTM, we can add unit tests in separate PR.

Leaving decision to @rnishtala-sumo

File: pkg/exporter/syslogexporter/exporter.go (hunk 5 above)
Comment (rnishtala-sumo) [reply pkg/exporter/syslogexporter/exporter.go:117]: will add the test for this in a new PR

**Pull request merged**